source/js/common/inject-react-components.js

```diff
@@ -0,0 +1,41 @@
+import React from "react";
+import ReactDOM from "react-dom";
+import JoinUs from "../components/join/join.jsx";
+
+/**
+ * Inject newsletter signup forms
+ * @param {Array} apps The existing array we are using to to track all ReactDOM.render calls
+ * @param {String} siteUrl Foundation site base URL
+ * @param {String} csrfToken CSRF Token
+ */
+const injectJoinUs = (apps, siteUrl, csrfToken) => {
+  // excluding `.join-us.on-nav` because it's taken care of by nav-newsletter.js
+  document.querySelectorAll(`.join-us:not(.on-nav)`).forEach(element => {
+    const props = element.dataset;
+    const sid = props.signupId || 0;
+
+    props.apiUrl = `${siteUrl}/api/campaign/signups/${sid}/`;
+
+    props.csrfToken = props.csrfToken || csrfToken;
+    props.isHidden = false;
+
+    apps.push(
+      new Promise(resolve => {
+        ReactDOM.render(
+          <JoinUs {...props} whenLoaded={() => resolve()} />,
+          element
+        );
+      })
+    );
+  });
+};
+
+/**
+ * Inject React components
+ * @param {Array} apps The existing array we are using to to track all ReactDOM.render calls
+ * @param {String} siteUrl Foundation site base URL
+ * @param {String} csrfToken CSRF Token
+ */
+export default (apps, siteUrl, csrfToken) => {
+  injectJoinUs(apps, siteUrl, csrfToken);
```

Comment:
I take it the idea is to continue building this part out so that main.js/bg-main.js become nice and tiny? If so, it might be worth already pre-splitting this up into this file, and a separate `./common/inject-react/join-us.js`, because there are a lot of other things we inject that should move out of the `main` files given this refactor.